IdentityCore/src/cache/MSIDMacKeychainTokenCache.m

```diff
@@ -0,0 +1,660 @@
+                                     // Since the attribute is an NSNumber, hash the string.
+                                     (id)kSecAttrCreator: [NSNumber numberWithUnsignedInt:(uint32_t)self.keychainGroup.hash],
+
+                                     // Add a marker for all cache items in the keychain for additional query filtering.
+                                     (id)kSecAttrLabel: s_defaultKeychainLabel
+
+                                     };
+
+        MSID_LOG_INFO(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", [self keychainGroupLoggingName]);
+        MSID_LOG_INFO_PII(nil, @"Init MSIDMacKeychainTokenCache with keychainGroup: %@", self.keychainGroup);
+    }
+
+    return self;
+}
+
+#pragma mark - Accounts
+
+// Write an account to the macOS keychain cache.
+//
+// Errors:
+// * MSIDErrorDomain/MSIDErrorInternal: invalid key or json deserialization error
+// * MSIDKeychainErrorDomain/OSStatus: Apple status codes from SecItemUpdate()/SecItemAdd()
+//
+- (BOOL)saveAccount:(MSIDAccountCacheItem *)account
+                key:(MSIDCacheKey *)key
+         serializer:(id<MSIDAccountItemSerializer>)serializer
+            context:(id<MSIDRequestContext>)context
+              error:(NSError **)error
+{
+    MSID_TRACE;
+    MSID_LOG_VERBOSE(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", _PII_NULLIFY(key.account), _PII_NULLIFY(key.service), [self keychainGroupLoggingName]);
+    MSID_LOG_VERBOSE_PII(context, @"Set keychain item, key info (account: %@ service: %@, keychainGroup: %@)", key.account, key.service, self.keychainGroup);
+
+    if (!key.service.length)
+    {
+        [self createError:@"Set keychain item with invalid key (service is nil)."
+                domain:MSIDErrorDomain errorCode:(NSInteger)MSIDErrorInternal error:error context:context];
+        return FALSE;
+    }
+
+    if (!key.account.length)
+    {
+        [self createError:@"Set keychain item with invalid key (account is nil)."
+                   domain:MSIDErrorDomain errorCode:(NSInteger)MSIDErrorInternal error:error context:context];
+        return FALSE;
+    }
+
+    NSData *jsonData = [serializer serializeAccountCacheItem:account];
+    if (!jsonData)
+    {
+        [self createError:@"Failed to serialize account to json data."
+                   domain:MSIDErrorDomain errorCode:(NSInteger)MSIDErrorInternal error:error context:context];
+        return FALSE;
+    }
+
+    NSMutableDictionary *query = [self accountQueryForKey:key];
+    NSMutableDictionary *update = [self accountUpdateForKey:key];
+    update[(id)kSecValueData] = jsonData;
+    OSStatus status = SecItemUpdate((CFDictionaryRef)query, (CFDictionaryRef)update);
+    MSID_LOG_INFO(context, @"Keychain update status: %d", (int)status);
+
+    if (status == errSecItemNotFound)
+    {
+        [query addEntriesFromDictionary:update];
```

Comment:
nit: there's only a need to create update dictionary and set data into it inside this "if-condition"

Reply:
It's also used above in the `SecItemUpdate()` call.